### DISCUSSION TITLE
[Feature]: Expose `errorInfo` from the `ErrorBoundary`

### DISCUSSION THREAD

**Discussion by marvinroger** [remix-run/react-router#9881] [Proposals]
### What is the new or updated feature that you are suggesting?

A React [`ErrorBoundary` `componentDidCatch`  exposes two things](https://reactjs.org/docs/react-component.html#componentdidcatch):

* `error`, which is the thrown error
* `info`, "An object with a componentStack key containing [information about which component threw the error](https://reactjs.org/docs/error-boundaries.html#component-stack-traces)"

My understanding is that React Router implements an `ErrorBoundary` and forwards the error to `errorElement`, which can get the error by using `useRouterError`. However, only the `error` is accessible, not the `info`.

My feature request is the following: being able to access, from within the `errorElement`, to the `info`.
I understand that `getDerivedStateFromError` does not expose `info`, so it might tricky to access it from the `errorElement`. In this case, an `onError(error, info)` callback on `RouterProvider` might also be a solution?

### Why should this feature be included?

Currently, we lose this `info` and, to my knowledge, there's no way to get it back, as we cannot create our own React `ErrorBoundary`, given the React Router one takes precedence. Am I wrong?

**Comment by ha404** [comment]
I'm having issue not able to wrap a specific component with a custom `ErrorBoundary` :(

**Comment by abetoots** [comment]
Having the same issue. Would be great to have access to the componentStack

**Comment by brophdawg11** [comment]
Duplicate proposal closed in favor of this one: https://github.com/remix-run/react-router/discussions/11206

**Reply by brophdawg11** [reply]
Cross-linking another related proposal: https://github.com/remix-run/react-router/discussions/10166

**Comment by brophdawg11** [comment]
> I understand that getDerivedStateFromError does not expose info, so it might tricky to access it from the errorElement. In this case, an onError(error, info) callback on RouterProvider might also be a solution?

Yeah this is where we've landed in our discussions around this,  I think we'll be looking into landing some form of API to let you log this information out/send it to Sentry, etc. - but you won't have it in the `ErrorBoundary` via `useRouteError` for 2 reasons:

* Logging is the documented purpose of [`componentDidCatch`](https://react.dev/reference/react/Component#componentdidcatch):
  * _"this lets you log that error to an error reporting service in production"_
* Setting state is deprecated in [`componentDidCatch`](https://react.dev/reference/react/Component#componentdidcatch)
  * _"In the past, it was common to call setState inside componentDidCatch in order to update the UI and display the fallback error message. This is deprecated in favor of defining [static getDerivedStateFromError.](https://react.dev/reference/react/Component#static-getderivedstatefromerror)"_

Changes are this will basically be a client-side version of the Remix [`handleError`](https://remix.run/docs/en/main/file-conventions/entry.server#handleerror) API for logging out error son the server.

**Reply by Tasin5541** [reply]
This would render components exposed through various analytics libraries useless and require users to basically build the same thing manually through the callback approach.

**Comment by wizardlyhel** [comment]
We have similar request but it's more around how to surface the partial failed deferred promise errors. For example, I have the following loader:

```tsx
export async function loader() {
  return {
    promise1: new Promise((resolve, reject) => setTimeout(() => {
      resolve('stuff');
    }, 1000)),
    promise2: new Promise((resolve, reject) => setTimeout(() => {
      reject(new Response('failed', {status: 404}));
    }, 3000)),
  };
}
```

It is intentional that I reject a promise with `Response`. This caused the `.data` payload to return a `SanitizedError` message.

![Screenshot 2025-01-17 at 10 07 46 AM](https://github.com/user-attachments/assets/c5ef6c34-4d93-4d3a-b02a-7b35caca942d)

However, there is no way to know that a failed deferred promise happened unless we `<Await>` on the failed promise and defined an `errorElement` on the `<Await>` component and manually log out an error .. and even if we did, in this case, a `SanitizedError` would make no sense.

So maybe ... a hook to log a failed deferred promise on the server-side, with information about a deferred Promise has returned a Response - which is not a proper data format. This is just for this specific example. But it would be nice to know why a deferred promise failed instead of going silent.

**Comment by epn-synth** [comment]
Hello! What's the status of this request? Is there any blocker preventing the implementation of the `onError` callback?

**Comment by brophdawg11** [comment]
We are going to implement this as a `<RouterProvider onError>` prop - tracking in https://github.com/remix-run/react-router/issues/12958.

**Comment by brophdawg11** [comment]
This is available in an alpha release via #14162 (`0.0.0-experimental-2eb7164`) if anyone wants to give it a spin

**Reply by rossipedia** [reply]
omgosh I was gonna propose this today. One step ahead ✊

**Comment by brophdawg11** [comment]
This was stabilized in 7.11.0
